components/form-control/script.js

```diff
@@ -0,0 +1,313 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Coherent Labs AD. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import components from 'coherent-gameface-components';
+import 'url-search-params-polyfill';
+
+const formMethods = {
+    GET: 'GET',
+    POST: 'POST'
+};
+
+const tags = {
+    BUTTON: 'button',
+    INPUT: 'input',
+    TEXTAREA: 'textarea',
+    GAMEFACE_CHECKBOX: 'gameface-checkbox',
+    GAMEFACE_DROPDOWN: 'gameface-dropdown',
+    GAMEFACE_RADIO_GROUP: 'gameface-radio-group',
+    GAMEFACE_RANGESLIDER: 'gameface-rangeslider',
+    GAMEFACE_SWITCH: 'gameface-switch'
+}
+
+const VALID_SUBMIT_ELEMENT_TAGS = new Set([tags.BUTTON, tags.INPUT]);
+const VALID_FORM_CONTROL_ELEMENT_TAGS = new Set([tags.BUTTON, tags.INPUT, tags.TEXTAREA]);
+const VALID_FORM_CONTROL_CUSTOM_ELEMENT_TAGS = new Set([
+    tags.GAMEFACE_CHECKBOX,
+    tags.GAMEFACE_DROPDOWN,
+    tags.GAMEFACE_RADIO_GROUP,
+    tags.GAMEFACE_RANGESLIDER,
+    tags.GAMEFACE_SWITCH
+]);
+
+class GamefaceFormControl extends HTMLElement {
+    constructor() {
+        super();
+        this.currentSubmitButton = null;
+    }
+
+    get method() {
+        return this.getAttribute('method') || formMethods.GET;
+    }
+
+    get action() {
+        return this.getAttribute('action') || '';
+    }
+
+    /**
+     * Will serialize the form data by traversing all the form element tree
+     * @returns {string}
+     */
+    getFormSerializedData() {
+        const params = new URLSearchParams();
+
+        //Serialize the data if there is any set on the submit button
+        this.serializeSubmitButtonElementData(params);
+        this.traverseFormElements(this, (element) => {
+            //We dont need data from invalid elements and elements that are having type "submit"
+            if (!this.isValidFromElement(element) || this.isElementUsedToSubmit(element)) return;
+
+            this.serializeElementData(element, params);
+        })
+
+        return params.toString();
+    }
+
+    /**
+     * Will serialize the data from the submit button if it has value attribute and it is <button> or <input type="submit"/>.
+     * @param {URLSearchParams} params
+     */
+    serializeSubmitButtonElementData(params) {
+        if (!this.currentSubmitButton) return;
+
+        const tagName = this.currentSubmitButton.tagName.toLowerCase();
+        if (tagName !== tags.BUTTON || tagName !== tags.INPUT) return;
```

Comment:
You are totally right about the check.
About the set you are not right. Making a check in set is O(1) and if I change it to array and check with indexOf then complexity will raise to O(n) wich is not a big deal in this case because we have two elements. Also sets are good for a static list no matter of the list length instead if the length is equal to 1.. :D 
I prefer the approach with the Set, but if you insist I will change it to array, there are no problems :D

Reply:
Set is fine, I thought that the check with an array will be easier in terms of code length, but didn't think about the complexity of execution.